src/Surfnet/StepupGateway/U2fVerificationBundle/Tests/Service/VerificationServiceAuthenticationVerificationTest.php

```diff
@@ -0,0 +1,123 @@
+        $registration = m::mock(new Registration(new KeyHandle($keyHandle), new PublicKey($publicKey)));
+        $registration->shouldReceive('authenticationWasVerified')->once()->with($updatedSignCounter);
+
+        $registrationDtoBeforeVerification = new RegistrationDto();
+        $registrationDtoBeforeVerification->keyHandle   = $keyHandle;
+        $registrationDtoBeforeVerification->publicKey   = $publicKey;
+        $registrationDtoBeforeVerification->signCounter = 0;
+
+        $registrationDtoAfterVerification = clone $registrationDtoBeforeVerification;
+        $registrationDtoAfterVerification->signCounter = $updatedSignCounter;
+
+        $registrationRepository = m::mock('Surfnet\StepupGateway\U2fVerificationBundle\Repository\RegistrationRepository');
+        $registrationRepository
+            ->shouldReceive('findByKeyHandle')
+            ->with(m::anyOf(new KeyHandle($keyHandle)))
+            ->andReturn($registration);
+        $registrationRepository
+            ->shouldReceive('save')
+            ->once()
+            ->with($registration);
+
+        $u2fService = m::mock('Surfnet\StepupU2fBundle\Service\U2fService');
+        $u2fService
+            ->shouldReceive('verifyAuthentication')
+            ->with(m::anyOf($registrationDtoBeforeVerification), m::anyOf($request), m::anyOf($response))
+            ->andReturn(AuthenticationVerificationResult::success($registrationDtoAfterVerification));
+
+        $service = new VerificationService($registrationRepository, $u2fService);
+        $service->verifyAuthentication($request, $response);
+    }
+
+    /**
+     * @test
+     * @group registration
+     */
+    public function it_doesnt_save_the_registration_on_failed_authentication()
+    {
+        $keyHandle = 'key-handle';
+        $publicKey = 'public-key';
+
+        $request = new SignRequest();
+        $request->keyHandle = $keyHandle;
+
+        $response = new SignResponse();
+
+        $registration = m::mock(new Registration(new KeyHandle($keyHandle), new PublicKey($publicKey)));
+        $registration->shouldReceive('authenticationWasVerified')->never();
+
+        $registrationDto = new RegistrationDto();
+        $registrationDto->keyHandle   = $keyHandle;
+        $registrationDto->publicKey   = $publicKey;
+        $registrationDto->signCounter = 0;
+
+        $registrationRepository = m::mock('Surfnet\StepupGateway\U2fVerificationBundle\Repository\RegistrationRepository');
+        $registrationRepository
+            ->shouldReceive('findByKeyHandle')
+            ->once()
+            ->with(m::anyOf(new KeyHandle($keyHandle)))
+            ->andReturn($registration);
+        $registrationRepository->shouldReceive('save')->never();
+
+        $u2fService = m::mock('Surfnet\StepupU2fBundle\Service\U2fService');
+        $u2fService
+            ->shouldReceive('verifyAuthentication')
+            ->with(m::anyOf($registrationDto), m::anyOf($request), m::anyOf($response))
+            ->once()
```

Comment:
I really wouldn't test for whether or not the u2fService has this method invoked with any parameters. What is being tested is that if this service returns a failed response a method in another class is not invoked. That is the only expectation that should be present, otherwise there is more than 1 unit under test. 

Allmost rieks as if Behat tests would be more appropriate: given a certain request and a certain DB state, a certain response is and certain DB state are expected. Much bigger than just testing a single unit, you're testing the boundary interactions it seems. Thoughts?